### PR TITLE
add initReady and init method in framework

### PR DESCRIPTION
**Pull request by MouceL**
slime supports getting configuration from xdsserver.

if this xdsserver is part of a slime submodule in bundle, it appears to call itself, framework connect xds configsource may failed.

so we should support starting up submodules  even if xds source not ready,

```
configSources:
- address: 'xds://localhost:16010'
```

### CODE REVIEW

**Comment by YonkaFang** [comment]
这个太反模式了，framework不应当假定依赖module状态。

File: framework/bootstrap/configcontroller.go [hunk 1]
```diff
@@ -357,38 +352,56 @@ func (c *configController) RegisterEventHandler(kind resource.GroupVersionKind,
...
 	<-stop
-	// stop k8s informers
-	c.stop <- struct{}{}
+	log.Infof("stop controller run")
+}
+
+func (c *Controller) Ready() bool {
+	for _, controller := range c.monitorControllers {
+		if !controller.Ready() {
+			log.Infof("config %+v not ready", controller.kind)
```
Comment (believening) [inline framework/bootstrap/configcontroller.go:366]: `Ready` just tell the caller if the `Controller` is ready. In my opinion, which `monitorController` is not ready is more suitable to be logged as debug information.

File: framework/bootstrap/configcontroller.go [hunk 2]
```diff
@@ -35,70 +35,64 @@ type ConfigController interface {
 	viewstore.ViewerStore
 	RegisterEventHandler(kind resource.GroupVersionKind, handler func(resource.Config, resource.Config, Event))
 	Run(stop <-chan struct{})
+	// Ready return all controller has ready
```
Comment (believening) [inline framework/bootstrap/configcontroller.go:38]: It is confusing to have `all controller` in the `ConfigController` interface method comment, and it is better for me to only reflect the state of this `ConfigController` itself.

File: framework/bootstrap/configcontroller.go [hunk 3]
```diff
@@ -495,21 +508,46 @@ func startXdsMonitorController(mc *monitorController, configRevision string, xds
...
-	// check xds cache synced or not
-	log.Infof("syncing data of xds config source [%s]", mc.configSource.Address)
-	for {
-		if xdsMCP.HasSynced() {
-			break
+	xdsMCP.Run()
+
+	go func(stop <-chan struct{}) {
+		log.Infof("syncing data of xds config source [%s]", mc.configSource.Address)
+		first := make(chan interface{}, 1)
```
Comment (believening) [inline framework/bootstrap/configcontroller.go:515]: can just use `chan struct{}`

File: framework/bootstrap/configcontroller.go [hunk 4]
```diff
@@ -495,21 +508,46 @@ func startXdsMonitorController(mc *monitorController, configRevision string, xds
...
+		first <- struct{}{}
+		ticker := time.NewTicker(1 * time.Second)
+		defer ticker.Stop()
+		timer := time.NewTimer(time.Duration(configSourceTimeout) * time.Second)
+		for {
+			select {
+			case <-first:
+				log.Infof("first time to connect to xds source")
+				if hasSync(xdsMCP, mc) {
```
Comment (believening) [inline framework/bootstrap/configcontroller.go:524]: can we move this out of the `select`?

File: framework/bootstrap/monitorcontroller.go [hunk 5]
```diff
@@ -4,6 +4,7 @@ import (
 	"errors"
 	log "github.com/sirupsen/logrus"
 	bootconfig "slime.io/slime/framework/apis/config/v1alpha1"
+	"sync"
```
Comment (believening) [inline framework/bootstrap/monitorcontroller.go:7]: `goimport` is a  good linter to organize the imports.

File: framework/bootstrap/monitorcontroller.go [hunk 6]
```diff
@@ -13,6 +14,8 @@ type monitorController struct {
 	configStore  ConfigStore
 	kind         monitorControllerKind
 	configSource *bootconfig.ConfigSource
+	ready        bool
+	sync.Mutex
```
Comment (believening) [inline framework/bootstrap/monitorcontroller.go:18]: `RMutex`

File: framework/model/module/module.go [hunk 7]
```diff
@@ -327,15 +327,26 @@ func Main(bundle string, modules []Module) {
...
+			return
+		}
 	}
 
-	var once sync.Once
-	ctx, cancel := context.WithCancel(ctrl.SetupSignalHandler())
+	go func() {
+		<-timeout
+		log.Errorf("timeout for config controller %s", config.Global.Misc["configsourceTimeout"])
+		cancel()
```
Comment (believening) [inline framework/model/module/module.go:348]: once

File: staging/src/slime.io/slime/modules/limiter/controllers/envoy_filters.go [hunk 8]
```diff
@@ -239,16 +239,20 @@ func getIstioServiceLabels(svc *slime_serviceregistry.Service) map[string]string
 
 func getIstioService(r *SmartLimiterReconciler, nn types.NamespacedName) (*slime_serviceregistry.Service, error) {
 	// get config from framework
-	config := r.env.ConfigController.Get(resource.IstioService, nn.Name, nn.Namespace)
-	if config == nil {
-		return nil, fmt.Errorf("get empty config base on %s/%s", nn.Namespace, nn.Name)
-	}
+	if r.env.ConfigController.Ready() {
```
Comment (believening) [inline staging/src/slime.io/slime/modules/limiter/controllers/envoy_filters.go:242]: We don't need to check ready every time

File: framework/bootstrap/configcontroller.go [hunk 9]
```diff
@@ -357,38 +353,54 @@ func (c *configController) RegisterEventHandler(kind resource.GroupVersionKind,
...
 	mc.configSource = configSource
 	log.Infof("init xds config source [%s] successfully", configSource.Address)
-	return mc, nil
+	return mc
 }
 
-func startXdsMonitorController(mc *monitorController, configRevision string, xdsSourceEnableIncPush bool) error {
+func startXdsMonitorController(mc *monitorController, config *bootconfig.Config, stopCh <-chan struct{}, timeout chan<- struct{}) error {
+	configRevision := config.Global.ConfigRev
+	xdsSourceEnableIncPush := config.Global.Misc["xdsSourceEnableIncPush"] == "true"
```
Comment (YonkaFang) [inline framework/bootstrap/configcontroller.go:392]: 所有的参数应该有说明，起码新增的。

File: framework/bootstrap/configcontroller.go [hunk 10]
```diff
@@ -357,38 +353,54 @@ func (c *configController) RegisterEventHandler(kind resource.GroupVersionKind,
...
+	configSourceTimeout, err := strconv.Atoi(config.Global.Misc["configsourceTimeout"])
+	if err != nil {
+		log.Errorf("parse configsourceTimeout %d err", configSourceTimeout)
```
Comment (YonkaFang) [inline framework/bootstrap/configcontroller.go:395]: 参数错误 与 参数没设置 不一样，前者一般不走到default。

File: framework/bootstrap/configcontroller.go [hunk 11]
```diff
@@ -495,21 +507,43 @@ func startXdsMonitorController(mc *monitorController, configRevision string, xds
 			}
 		},
 	})
+	xdsMCP.Run()
 
-	if err = xdsMCP.Run(); err != nil {
-		return fmt.Errorf("failed to init client for xds config source %s, err: %+v", mc.configSource.Address, err)
-	}
+	go func(stop <-chan struct{}) {
+		log.Infof("syncing data of xds config source [%s]", mc.configSource.Address)
```
Comment (YonkaFang) [inline framework/bootstrap/configcontroller.go:513]: 这一段可以简化一下，写的较为复杂

File: framework/bootstrap/configcontroller.go [hunk 12]
```diff
@@ -495,21 +507,43 @@ func startXdsMonitorController(mc *monitorController, configRevision string, xds
...
+		}
 
-	// check xds cache synced or not
-	log.Infof("syncing data of xds config source [%s]", mc.configSource.Address)
-	for {
-		if xdsMCP.HasSynced() {
-			break
+		ticker := time.NewTicker(1 * time.Second)
+		defer ticker.Stop()
+		timer := time.NewTimer(time.Duration(configSourceTimeout) * time.Second)
```
Comment (YonkaFang) [inline framework/bootstrap/configcontroller.go:520]: 接受 wait-forever-until 语义吗？

File: framework/model/module/module.go [hunk 13]
```diff
@@ -327,15 +327,30 @@ func Main(bundle string, modules []Module) {
...
+		if err != nil {
+			log.Errorf("run config controller failed: %s", err)
+			return
+		}
 	}
 
-	var once sync.Once
-	ctx, cancel := context.WithCancel(ctrl.SetupSignalHandler())
+	go func() {
+		<-timeout
```
Comment (YonkaFang) [inline framework/model/module/module.go:350]: 我还是认为，用timeout来解决依赖问题不是合适的方式。 一些场景下就是刚性依赖。

**Pull request closed without merging**